### PR TITLE
Upgrade to golang 1.6

### PR DESCRIPTION
**Pull request by luxas**
A second attempt to upgrade go version above `go1.4`
Merge ASAP after you've cut the `release-1.2` branch and feel ready.
`go1.6` should perform slightly better than `go1.5`, so this time it might work
@gmarek @wojtek-t @zmerlynn @mikedanese @brendandburns @ixdy @thockin

### CODE REVIEW

**Comment by luxas** [comment]
@mikedanese Should we have an option to force build the `kube-cross` image when necessary? 
In these cases that's needed.

File: build/build-image/cross/Dockerfile [hunk 1]
```diff
@@ -15,7 +15,7 @@
 # This file creates a standard build environment for building cross
 # platform go binary for the architecture kubernetes cares about.
 
-FROM  golang:1.4.2
+FROM golang:1.6
```
Comment (mikedanese) [inline build/build-image/cross/Dockerfile]: please include the minor version. e.g. 1.6.0

**Comment by mikedanese** [comment]
.travis.yml and shippable.yml also need the version bump

**Comment by mikedanese** [comment]
> Should we have an option to force build the kube-cross image when necessary? 
> In these cases that's needed.

What would it do? We need to push kube-cross before CI goes green. That seems like a reasonable check to have

**Comment by luxas** [comment]
@mikedanese Fixed. Now the `kube-cross` will be built if it isn't pushed.

**Comment by mikedanese** [comment]
I don't think that's a good idea. It allows CI to pass without pushing the image first which is a bad thing.

**Comment by luxas** [comment]
I'm fine with that. Do you have a better approach?

**Comment by luxas** [comment]
@mikedanese OK, what do you say about this then? 
We'll update non-release CI to `go1.6` in this PR before `v1.2`.
If you want, we may also merge the bump to `go1.6` in `build/build-image/cross` before `v1.2`
If we do that, you should push `v1.6.0-1` to `gcr.io`, **but we'll not use it in `build/common.sh` for `v1.2`**

If a `v1.6.0-1` image would be pushed somehow, it would help a lot!

**Comment by wojtek-t** [comment]
Well - we deinifetly do not want to have this for 1.2, so I will get back to it once we are done with our 1.2 work.

**Comment by luxas** [comment]
Yeah, as I said in the first place. Didn't think you'd want it either. But just asked anyway...
I'm ready when you are ready :)

File: hack/jenkins/test-image/Dockerfile [hunk 2]
```diff
@@ -15,7 +15,7 @@
 # This file creates a build environment for building and running kubernetes
 # unit and integration tests
 
-FROM golang:1.5.3
```
Comment (ixdy) [inline hack/jenkins/test-image/Dockerfile]: should update the tag for this image (in the Makefile) as well

Comment (luxas) [inline hack/jenkins/test-image/Dockerfile]: Thanks. Fixed.

**Comment by wojtek-t** [comment]
@luxas - FYI, we have a bunch of unit tests failing with Go 1.6, so there is long way to do before we do this...

**Comment by luxas** [comment]
Thanks for the info, good to know. 
But really sad things are failing.
Would intermediate `1.4.2` => `1.5.3` => `1.6.0` be a better migration, so the step wouldn't be so huge?
Just thinking out loud...

**Comment by mikedanese** [comment]
@luxas we already upgraded to 1.5.3 and had to roll back because our scalability tests started failing.

**Comment by mikedanese** [comment]
We also rolled back from 1.4.3

**Comment by ixdy** [comment]
@luxas @mikedanese x-ref #20656

**Comment by luxas** [comment]
Yeah, I know that we upgraded to `go1.5.3`, cause I made the PR :) #19985

**Comment by luxas** [comment]
I've also read the issues and such, but I meant that `go1.5.3` might be a good intermediate step for `v1.3`, so you may investigate the scalability issues like #20400 if `go1.6` is even worse

**Comment by luxas** [comment]
Rebased

**Comment by spxtr** [comment]
It's likely some of the e2e errors are because of the go version. Lets re-run e2e to make sure.

@k8s-bot e2e test this again issue: #IGNORE

**Comment by spxtr** [comment]
Second pass had several failures as well:

```
[Fail] [k8s.io] Port forwarding [k8s.io] With a server that expects no client request [It] should support a client that connects, sends no data, and disconnects [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:249

[Fail] [k8s.io] Kubectl client [k8s.io] Simple pod [It] should support port-forward 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:520

[Fail] [k8s.io] Port forwarding [k8s.io] With a server that expects a client request [It] should support a client that connects, sends no data, and disconnects [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:142

[Fail] [k8s.io] Port forwarding [k8s.io] With a server that expects a client request [It] should support a client that connects, sends data, and disconnects [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:193
```

**Comment by luxas** [comment]
@spxtr The `e2e` failures are expected, cause `kube-cross` v1.6.0-1 hasn't been pushed yet.

**Comment by spxtr** [comment]
It's pretty interesting that this set of tests fails here.

**Comment by spxtr** [comment]
```
[It] should support port-forward
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:525
STEP: forwarding the container port to a local port
Mar 17 13:29:06.694: INFO: starting port-forward command and streaming output
Mar 17 13:29:06.694: INFO: Asynchronously running '/var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce@2/kubernetes/platforms/linux/amd64/kubectl kubectl --server=https://104.197.179.216 --kubeconfig=/var/lib/jenkins/workspace/kubernetes-pull-build-test-e2e-gce@2/.kube/config port-forward --namespace=e2e-tests-kubectl-l5lx9 nginx :80'
Mar 17 13:29:06.705: INFO: reading from `kubectl port-forward` command's stderr
STEP: curling local port output
Mar 17 13:29:07.305: INFO: got: 
Mar 17 13:29:07.305: FAIL: Failed http.Get of forwarded port (http://localhost:53120): Get http://localhost:53120: read tcp 127.0.0.1:48104->127.0.0.1:53120: read: connection reset by peer
```

**Comment by spxtr** [comment]
The port-forwarding e2es are actually broken, I verified on my local machine with k8s from this PR.

**Comment by ncdc** [comment]
I've got a Fedora Rawhide VM with go1.6, but unfortunately the version of Docker I have on it isn't working at the moment. Once I get that resolved, I can look at the port forwarding failures.

**Comment by ncdc** [comment]
I've got Docker working now, so I can look at the test failures.

**Comment by ncdc** [comment]
Figured out the port forwarding failures. Apparently go 1.6 doesn't like this: https://github.com/kubernetes/kubernetes/blob/9a71dfe753b4e67016f306ac18d6a27698f4517b/test/e2e/portforward.go#L90-L91. I'm guessing when you close stdout and/or stderr prematurely, maybe it's terminating the `kubectl port-forward` process too soon?

**Comment by spxtr** [comment]
Might be related to [this change](https://go-review.googlesource.com/#/c/11415/).

> In the os/exec package, Cmd's Output method continues to return an ExitError when a command exits with an unsuccessful status. If standard error would otherwise have been discarded, the returned ExitError now holds a prefix and suffix (currently 32 kB) of the failed command's standard error output, for debugging or for inclusion in error messages. The ExitError's String method does not show the captured standard error; programs must retrieve it from the data structure separately.

**Comment by ncdc** [comment]
stdout/stderr in this case are pipes (StdoutPipe/StderrPipe). Not sure if that matters.

**Comment by rsc** [comment]
@ncdc, the test/e2e directory contains this pattern a few times:

	cmd := ... *os/exec.Cmd ...
	stdout, stderr, err := startCmdAndStreamOutput(cmd)
	defer stdout.Close()
	defer stderr.Close()
	...
	stderr.Read(buf)
	...
	return cmd

As written, this code connects a pipe to a new process's standard error, reads some output from that pipe, closes that pipe, and then expects the command to continue running. The new process will continue running until it tries to write to standard error again, at which point the write to the closed pipe will cause the new process to get a SIGPIPE. By default, Unix programs die on SIGPIPE. 

Sadly, SIGPIPE is now also sent when writing to a closed network connection. It makes sense to kill a program writing to a closed pipe in a pipeline: if you're running `x | y | z` and y dies, it usually doesn't make sense for x to keep running. But it does not make nearly so much sense to kill a program writing to a closed TCP connection. If a web server's client hangs up early, the web server shouldn't be killed.

To try to do the right thing in the networked context, Go programs have historically not died on SIGPIPE, at least not immediately. Instead, they ignored SIGPIPE, which meant the offending write would get an EPIPE error, and then the write treated EPIPE like any other error until it happened too many times in a row (to be precise, I think it was 10), and then they died. This confused basically everyone. And the behavior was turned off entirely for network connections, but it still applied to things like running subprocesses (for example cgi).

Go 1.6 tried to unconfuse things by going back to something more like Unix behavior. A SIGPIPE writing to fd 1 or 2 kills the program immediately, like other Unix programs. And a SIGPIPE on any other fd is completely ignored (returns EPIPE no matter how many times it happens).

What I believe is going on in snippets like above is that the program being run as a subprocess is a Go program and writes a small amount to standard error at startup. The code snippet reads that amount and then closes standard error. Then later the subprocess writes a small amount more to standard error. Specifically, at least one write but fewer than 10 writes. If it had gone over 10 writes in Go 1.5 or earlier, it would have been killed even in the earlier releases. Now the cutoff is effectively 1 instead of 10, so the subprocess crosses the line earlier.

As far as I can tell, the fix is to change this pattern. Instead of deferring those closes you need something like

```
func drain(f *os.File) {
	go func() {
		io.Copy(ioutil.Discard, f)
		f.Close()
	}()
}
```

and then replace `defer stderr.Close()` with `defer drain(stderr)`. That will keep the pipe open and keep the subprocess from getting stuck (someone has to read the data out), until it's gone.

I count seven `defer stdout.Close()` and five `defer stderr.Close()` in the kubernetes git repo. All should be  checked.

**Comment by ncdc** [comment]
@rsc thanks for the detailed explanation!

**Comment by timothysc** [comment]
Do we have a canonical list of items here, I can help dig in on this one.

**Comment by ncdc** [comment]
@timothysc I think fixing the other instances mentioned in https://github.com/kubernetes/kubernetes/pull/22149#issuecomment-202508316 would be useful.

**Comment by spxtr** [comment]
@k8s-bot e2e test this issue: #23445

**Comment by luxas** [comment]
@spxtr If you think the `kube-cross:v1.6.0-1` Dockerfile looks fine, push it and we'll see what CI says

**Comment by spxtr** [comment]
That's curious, it used the 1.6 image for a few of the previous runs. I pushed the new version.

@k8s-bot e2e test this issue: #IGNORE

**Comment by spxtr** [comment]
Awesome.

**Comment by spxtr** [comment]
Lets merge on Monday.

**Comment by ixdy** [comment]
This is probably worthy of a release note, right?

**Comment by spxtr** [comment]
Probably.

**Comment by lavalamp** [comment]
@dchen1107

**Comment by hongchaodeng** [comment]
The PR LGTM.
Can we get this merged today?

**Comment by luxas** [comment]
Yes, I hope we may get this in today.

I guess @k8s-oncall should watch this closely.
One thing that probably is going to happen is longer build times, up to a factor of ~2.
That happened at least last time we tried to upgrade.

go1.5
> Builds in Go 1.5 will be slower by a factor of about two. ... Further profiling and optimization will continue in Go 1.6 and future releases.

go1.6
> On average the programs in the Go 1 benchmark suite run a few percent faster in Go 1.6 than they did in Go 1.5. ... There have been significant optimizations bringing more than 10% improvements to implementations of {some core packages}

**Comment by spxtr** [comment]
For the few builds I've looked at, it's going to go up from about 8 minutes to about 11 minutes for a full build-release on PR Jenkins.

**Comment by luxas** [comment]
@spxtr ok, good. That wasn't too much :)

File: shippable.yml [hunk 3]
```diff
@@ -4,7 +4,7 @@ language: go
 
 matrix:
   include:
-    - go: 1.5.3
+    - go: 1.6.0
```
Comment (wojtek-t) [inline shippable.yml]: @spxtr @ixdy 
Since we are no longer using shippable - can't we just remove this whole file?

File: docs/devel/development.md [hunk 4]
```diff
@@ -54,7 +54,7 @@ Kubernetes is written in the [Go](http://golang.org) programming language. If yo
 
 ### Go versions
 
-Requires Go version 1.4.x or 1.5.x
+Requires Go version 1.4.x, 1.5.x or 1.6.x
```
Comment (wojtek-t) [inline docs/devel/development.md]: To be honest - I think we do NOT officially support 1.5 Go (due to performance issues) - so we shouldn't mention it here.

Comment (hongchaodeng) [inline docs/devel/development.md]: If we run 1.6 in testing, it should require  >= 1.6.

Comment (wojtek-t) [inline docs/devel/development.md]: That's not entirely true - currently we are running 1.4.2 and we officially do NOT support 1.5.X (but it is >= 1.4.2).

We need to explicitly specify versions that we are supporting.

File: shippable.yml (hunk 3 above)
Comment (spxtr) [inline shippable.yml]: Yes, probably best to do that in a separate PR though.

Comment (wojtek-t) [inline shippable.yml]: https://github.com/kubernetes/kubernetes/pull/23845 out for review

**Comment by lavalamp** [comment]
LGTM

**Comment by timothysc** [comment]
Can someone apply the label, given multiple lgtms?

**Comment by eparis** [comment]
I'll be the one.   /me wonders why @timothysc didn't do it  :)

**Comment by timothysc** [comment]
@eparis didn't want to stomp on toes. ;-)

**Comment by eparis** [comment]
STOMP!

File: docs/devel/development.md (hunk 4 above)
Comment (luxas) [inline docs/devel/development.md]: I may fix that in a follow-up if you like

Comment (wojtek-t) [inline docs/devel/development.md]: @luxas - can you fix that in this PR? I will re-apply gtlm label as soon as this is done

**Comment by spxtr** [comment]
Here we go!

**Comment by hongchaodeng** [comment]
`./test/e2e_node/jenkins/e2e-node-jenkins.sh` seems to be using the `go` on host. This might still be using go 1.4 depending on the host.

@spxtr @luxas Is there any plan for keep that part up-to-date as well?

/cc @xiang90

File: docs/devel/development.md (hunk 4 above)
Comment (luxas) [inline docs/devel/development.md]: It merged already, will send a follow-up anyway
